New (additional) LCARS theme inspired by Picard TV show

MichalSvatos opened the pull request:
## What does this PR aim to accomplish?
New (additional) LCARS theme. Description and more screenshot available [here](https://github.com/MichalSvatos/pi-hole-star-trek-picard/). Related discussion on the [Discourse forum](https://discourse.pi-hole.net/t/star-trek-picard-lcars-theme-for-pi-hole/65003).

**Small preview**
![image](https://github.com/pi-hole/AdminLTE/assets/5910531/3c79f85a-5eea-4214-a8ca-dbbf84bb285a)

## How does this PR accomplish the above?
New CSS file added and configuration of the themes updated.

---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
6. I have checked that another pull request for this purpose does not exist.
7. I have considered, and confirmed that this submission will be valuable to others.
8. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
9. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

MichalSvatos committed:
Added new LCARS theme

Signed-off-by: Michal Svatos <github@svatos.dev>

rdwebdesign commented on style/themes/lcars-picard.css:
Use 2 spaces for indentation instead of tabs.

MichalSvatos replied:
Yes, I already see all the failed tests. Great start, I cannot say 😄 

(It's weird, because PHPStorm showed me 2 spaces are used.)

rdwebdesign replied:
I think there are tabs on the comments.

MichalSvatos commented:
I'm trying to solve another issue. I included the "Signed-by-off.." message, but it says [it's wrong](https://github.com/pi-hole/AdminLTE/pull/2709/checks?check_run_id=16941811595). But as far as I can see those 2 strings are identical 🤔 

`Expected "Michal Svatoš github@svatos.dev", but got "Michal Svatoš github@svatos.dev"`

rdwebdesign commented:
I don't know what the issue is.

Maybe it's an encoding difference (like [this one](https://github.com/dcoapp/app/issues/183)), or DCO test simply can't handle non-english strings (`Svatoš`).

MichalSvatos commented:
It crossed my mind, but I somehow presumed that a global company in 2023 could acount for that. I'll try to search more.

rdwebdesign commented:
I'm guessing about the encoding, but if I'm right, the name and encoding on your Github account should match the name (and encoding) set on your machine (`git config --global user.name`).

rdwebdesign commented:
The "Lint / PHP-CS-Fixer" error is unrelated to this PR.
We will probably fix it on another PR.

MichalSvatos commented:
Yes, I figure it out :slightly_smiling_face: I went after the prettier fix and the DCO, which was caused by the special characters as you pointed out previously :+1: 

Sorry, for the mess I had to setup everything for the GH signing and also I didn't realize (which I should have :facepalm: ) that the code style will be probably different from mine.

rdwebdesign commented:
I saw a few visual issues on desktop. Please, confirm if you see the same:

1. The gray background creating the curve is not correctly aligned (bottom part of the image):
  ![image](https://github.com/pi-hole/AdminLTE/assets/1385443/48f17627-0ab7-4649-8075-c1b3c591c27e).
  I fixed it adding `width: calc(var(--sidebar-width) - var(--gap25));` (line 1653).
  ![image](https://github.com/pi-hole/AdminLTE/assets/1385443/486901d8-7bfb-4539-b088-a45ebcec0c13)
  NOTE: this is happening only on the full width layout.

2. There is a secondary/internal scrollbar when using the full width layout:
  ![new_theme01](https://github.com/pi-hole/AdminLTE/assets/1385443/524a44fe-754a-4fbf-af40-0b16001af8bb)

3. Related to the same footer issue, the style you added to the footer bottom using `:after` selector is not visible when using boxed layout (it's hidden even when you scroll to bottom):
  ![image](https://github.com/pi-hole/AdminLTE/assets/1385443/582bbbe5-3942-48fe-9940-6b9effc5fd15)

MichalSvatos commented:
I see all the issues.

The first two could be solved by adding small padding to the wrapper `padding-inline: var(--gap25)` and `overflow: hidden`, but there is probably a reason why overflow-y is set to auto :thinking: 

The third one ... you're running it in the Docker I presume? Because on the "normal" version there are only 3 items in the footer and I put there the 4th one fill the visual hole and insert self promo at the same time :angel: I'll delete it.

At the same time I noticed another small issue with the footer.

Thank you for the feedback! I'll fix it during tomorrow.

yubiuser commented:
> We will probably fix it on another PR.

https://github.com/pi-hole/AdminLTE/pull/2710

yubiuser commented:
The notification alert is misaligned (should be attached to the arrows like in the other themes I guess?)
 
![Screenshot at 2023-09-20 13-47-22](https://github.com/pi-hole/AdminLTE/assets/26622301/6792e672-c88e-4847-b609-5a5a06eed633)

yubiuser commented:
Some glitches here as well
![Screenshot at 2023-09-20 13-49-27](https://github.com/pi-hole/AdminLTE/assets/26622301/7b2b9285-f6a0-4fb9-9a01-5e1724f50de0)
![Screenshot at 2023-09-20 13-50-44](https://github.com/pi-hole/AdminLTE/assets/26622301/00951f99-90b5-413c-86c1-087332b59a1d)

MichalSvatos commented:
> The notification alert is misaligned (should be attached to the arrows like in the other themes I guess?)
> 
> ![Screenshot at 2023-09-20 13-47-22](https://user-images.githubusercontent.com/26622301/269272207-6792e672-c88e-4847-b609-5a5a06eed633.png)

It should look like this:
<img width="147" alt="image" src="https://github.com/pi-hole/AdminLTE/assets/5910531/8e30850a-962f-4c44-a75b-a2f61bb14927">

MichalSvatos commented:
> Some glitches here as well
![Screenshot at 2023-09-20 13-49-27](https://user-images.githubusercontent.com/26622301/269272455-7b2b9285-f6a0-4fb9-9a01-5e1724f50de0.png)

Is this the same installation? To be honest I don't know how to fix this now from the top of my head :thinking: The theme needs the "Antonio" font. Or at least some "narrow" font. Any ideas?

(I think I will have to get the Docker up and running as soon as possible to be more flexible with the testing.)

yubiuser commented:
Yes, it's the same installation. I checked out your branch on my docker installation.

rdwebdesign commented:
I'm not seeing the same issue with the font:

![image](https://github.com/pi-hole/AdminLTE/assets/1385443/ff333917-9ca0-45bb-9e13-15f6398cbed8)

I will investigate.

rdwebdesign commented:
I don't want to interfere with your design choices, but I think the contrast is too low for the links on Hostname column (network table):

![image](https://github.com/pi-hole/AdminLTE/assets/1385443/c82b8e5e-3b6f-471f-8815-03b1db088c79)

Using something like this will improve the readability:
```css
table[id="network-entries"] td:nth-child(4) a:not(:hover) {
  color: #fff;
}
```
![image](https://github.com/pi-hole/AdminLTE/assets/1385443/1b2316b6-e037-411a-93d5-b6dc69b1ddd6)


What do you think?

MichalSvatos commented:
### Scrollbar + footer + rounded corners
The scrollbar was caused by footer (`:after` pseudoelement was too tall). I added the inline padding which fixed the rounded corners and it looks better. No `overflow: hidden` needed.

### Color in network table
You're right @rdwebdesign - fixed.
(Sidenote - the color palette used in that table is not ideal, but I didn't want to introduce completely new colors just for that. We can think about something later, it's not a deal breaker.)

### Number next to the arrow
I fixed the position to acount for bigger numbers, but the missalign in your screenshot is caused by the font again.

### Font
Can you check which kind of font is actually applied @yubiuser? I don't know why the "Antonio" font is not used. It's local, so it cannot be some blocker of 3rd party fonts. Maybe some custom settings?

(ℹ️ I rebased the branch to include PHP_CS)

rdwebdesign commented:
I found by accident (when I resized the browser window) another issue.

On the Dashboard, the legend for Query Types graphic (and probably also Upstream Servers graphic) is overflowing on certain conditions (big number of items and smaller screens, like tablets)

![image](https://github.com/pi-hole/AdminLTE/assets/1385443/8c9c16f2-7c9e-498d-a9ac-bdf6c90d47b4)

yubiuser commented:
> Can you check which kind of font is actually applied @yubiuser? 

Not sure how to check this. Inspection gave me
![Screenshot at 2023-09-20 22-21-16](https://github.com/pi-hole/AdminLTE/assets/26622301/2ab9f1a0-3474-4cef-a565-0ffb71bd0ebc)

MichalSvatos commented:
> On the Dashboard, the legend for Query Types graphic (and probably also Upstream Servers graphic) is overflowing on certain conditions (big number of items and smaller screens, like tablets)

Ouch. That's a problem 🤔 I didn't think there could be so much of them. But I knew that the position absolute was risky move.

I quickly tried several things (like calculating height based on viewport width and margins), but we will have to renounce the position absolute. It's not gonna look so great, but it will work.

What about something like this? (quick styling with dev tools)
<img width="675" alt="image" src="https://github.com/pi-hole/AdminLTE/assets/5910531/0a49446d-9400-4b92-acc4-8d0997e45df3">

rdwebdesign commented:
> What about something like this?

What will happen if there are 9 or 10 items?

MichalSvatos commented:
I'll have to tweak it little bit more, but this version is not based on `position: absolute`, so it will scale according to the number of items 🙂

MichalSvatos commented:
2 variants, both with dynamic height. Which one do you think is better @rdwebdesign?

![image](https://github.com/pi-hole/AdminLTE/assets/5910531/17a5d6be-05eb-47d5-94dd-af35c48dca04)

rdwebdesign commented:
> 2 variants, both with dynamic height. Which one do you think is better

I think it's your decision.

---

I found another small glitch. Now, there's a gap on mobile screens when you open the sidebar:

![image](https://github.com/pi-hole/AdminLTE/assets/1385443/cff490ce-8629-498a-b216-124968da5d4b)

I think this could be fixed adding:
```css
@media (max-width: 767px) {
  .sidebar-open .main-sidebar {
    transform: translate(var(--gap25), 0);
  }
}
```

MichalSvatos commented:
At the end I figure it out, how to keep the same design, but make it dynamic :wink: At the same time I slightly adjusted the mobile version as well.

Sidebar is fixed.

MichalSvatos commented:
- FIX - added min-height to `.content` to avoid incorrect footer position when TOOLS menu is open
- FIX - corrected position of the "timer" number in the right corner (boxed vs fluid layout)
- FIX - showed icons in the 4 status boxes on dashboard on mobile

(bumped version to 1.4.0)

rdwebdesign commented:
It's not forgotten.
I will review the PR when I find free time (be patient). 

I'm also requesting a review from the team.

MichalSvatos commented:
Sorry 🙂 Take your time and let me know if I can do anything.

yubiuser commented:
The issue I described above seems to be caused by the available space. The screencapture is with the latest version of this branch
![Peek 2023-10-06 23-23](https://github.com/pi-hole/AdminLTE/assets/26622301/59bee6f0-5050-4342-a3b1-b5c184e7dbe0)

rdwebdesign commented:
@yubiuser 

I'm not sure why, but your video is not using the expected font (Antonio).

![image](https://github.com/pi-hole/AdminLTE/assets/1385443/6eacfc33-13cf-4e12-928d-e069779d7168)

MichalSvatos commented:
Nor yours @rdwebdesign 🤔 Your is using "Myriad Pro Cond".

What I can do is at least add the left padding, overflow and text-overflow ... but it doesn't solve the problem with the font.

<img width="199" alt="image" src="https://github.com/pi-hole/AdminLTE/assets/5910531/2f5a2490-37e8-4168-b0b7-829c2f8e5125">

rdwebdesign commented:
I found the issue... 

PHP only loads the extra fonts for the original `lcars.css`:
https://github.com/pi-hole/AdminLTE/blob/41682f17b72d3fb83837a7a08fa68b3e37cd35b7/scripts/pi-hole/php/header.php#L44-L46

> Your is using "Myriad Pro Cond"

Yeah. I have this font installed on the machine I used to generate the image.

MichalSvatos commented:
@rdwebdesign 👏 Awesome! I didn't think of that at all. I'll fix it. Thank you!

MichalSvatos commented:
- updated condition for `lcars-picard`
- tweaked buttons (overflow + ellipsis) ... just in case 😉 

I'm so custom to see and work only the CSS file that I forgot that now I can/have to look at it as a whole project 🤦

Thank you again for the help debugging it!

rdwebdesign commented:
I noticed another thing.

You changed the font color and font weight for any Log output using this code:
https://github.com/pi-hole/AdminLTE/blob/6a4a147f438be8d5251d7caf2c98c3989a8c39d6/style/themes/lcars-picard.css#L5458-L5464

This is breaking the **_Debug Log_** colors, but the real issue is on the ***Tail pihole.log*** page.

This page is showing blocked queries and other queries in Green.
It's very hard to distinguish blocked queries:
![image](https://github.com/pi-hole/AdminLTE/assets/1385443/071f9f78-74f3-4034-af92-ed04f689f5a1)
 

On this page, other themes are following the same color pattern used by the command line:
- Default color (black on light themes, white on dark themes) + Bold for query lines;
- Red + bold for Blocked lines;
- dimmed color for everything else.

![image](https://github.com/pi-hole/AdminLTE/assets/1385443/0b81d741-78fb-4518-b0f4-faf4350824e1)

MichalSvatos commented:
That's a valid point. I'll look into it.

EDIT: Ok, so I've just discovered that the logs have also some classes 😆 eg `<span class="log-purple">*** [ DIAGNOSING ]:</span>`. I'll fix that (but maybe tomorrow). Thanks

rdwebdesign commented:
Yeah... every theme has these classes.

Examples:
- Light theme [log colors](https://github.com/pi-hole/AdminLTE/blob/41682f17b72d3fb83837a7a08fa68b3e37cd35b7/style/themes/default-light.css#L229-L250).
- LCARS [log colors](https://github.com/pi-hole/AdminLTE/blob/41682f17b72d3fb83837a7a08fa68b3e37cd35b7/style/themes/lcars.css#L2004-L2025).

MichalSvatos commented:
It was "lost in translation" apparently 🤦 I don't see it even in my TNG theme. 
Anyway - definitely fixing that 🙂

MichalSvatos commented:
Logs fixed. Anyway I had to add new colors.
<img width="389" alt="image" src="https://github.com/pi-hole/AdminLTE/assets/5910531/4a2e621f-f6ba-498b-999c-c67fcde059d4">
<img width="203" alt="image" src="https://github.com/pi-hole/AdminLTE/assets/5910531/03933387-9a78-4d23-adb6-4809bd0e445f">

yubiuser commented:
Font is sorted now, but there is some bar-text overlap on the footer
![Screenshot at 2023-10-08 08-39-26](https://github.com/pi-hole/AdminLTE/assets/26622301/1c276fef-6642-4318-bf42-6dfe480a6ec5)

MichalSvatos commented:
That's really weird 🤔 From the screenshot it looks like that the footer is completely broken. There must be some different HTML.

That's not a DEV version of docker, right? Because in my docker with `pihole/pihole-dev` it looks fine.

PromoFaux commented:
Something that is probably worth pointing out now, as I'm not sure it has already been pointed out...

We should probably better target `development-v6` with this change rather than `devel` - it's going to need to be converted in any case, so might as well do it now and release along with `v6`

MichalSvatos commented:
@PromoFaux Hi, `v6` was mentioned in the [original thread](https://discourse.pi-hole.net/t/star-trek-picard-lcars-theme-for-pi-hole/65003/21) on Discourse with the note that there is still not ETA for the release.

PromoFaux commented:
Ah, apologies - I had not seen that conversation on Discourse - Carry on, I said nothing!

(Though probably worth looking at what's needed to make it v6 compatible in the meantime)

yubiuser commented:
> That's not a DEV version of docker, right? Because in my docker with pihole/pihole-dev it looks fine.

Your're correct, it's the stable v5 image. I'll try it with the dev version

**ADD:** Using the dev container the footer looks fine

yubiuser commented:
The filter in `Domain Management` looks like this (on dev ;-)
![Screenshot at 2023-10-08 19-41-53](https://github.com/pi-hole/web/assets/26622301/ae2bc93e-3522-459c-8a56-ffd374255185)

Using boxed layout and having a long domain, the audit log will show de-arranged buttons

![Screenshot at 2023-10-08 19-48-33](https://github.com/pi-hole/web/assets/26622301/115c0a5d-1b86-4a00-a344-cca88a385c8d)

PromoFaux commented:
The :dev tag on docker still uses all the master branches of the other components. (It is designed to do docker Dev work against currently released pi-hole)

:nightly has all the Dev branches

MichalSvatos commented:
Audit log - got it 👍 

@PromoFaux So I should use `:nightly` to test the `devel` version? Or the `:nightly` tag also contains `v6` changes ...? 🤔 Sorry, for the lame question. I'm still struggling with the docker pi-hole.

PromoFaux commented:
the `nightly` tag contains the `development` branches for v5 (so core `development`, web `devel`, FTL `development`)

`:development-v6` contains all the `v6` branches (also named `development-v6` on all components)

MichalSvatos commented:
OIC. Thank you for the clarification 👍 

I see now the broken footer, btw. There is a different class `list-unstyled` vs `list-inline`.

MichalSvatos commented:
**1.4.2**
- FIX - checkboxes and radios for both `lcars` classes
- FIX - more general selector for footer styling `.list-inline` --> `ul[class*="list-"]`
- FIX - alert colors, border-radius and close button styling

rdwebdesign commented on style/themes/lcars-picard.css:
```diff
@@ -0,0 +1,5906 @@
+  ========================================================== */
+/* body[class*="lcars"][class*="lcars"] is a small dirty trick to avoid even more deep "!important hell" */
+
+body[class*="lcars"][class*="lcars"] [class*="icheck-"] > label {
```
Why did you add the attribute selector (`[class*="lcars"]`) twice?

```suggestion
body[class*="lcars"] [class*="icheck-"] > label {
```

MichalSvatos replied:
I made a comment there to clarify -  `/* body[class*="lcars"][class*="lcars"] is a small dirty trick to avoid even more deep "!important hell" */`

The `pihole.css` contains several lines of code with `body:not(.lcars) ...`. `:not` adds specificity, so I needed to do the same, but without depending on another class or putting !important everywhere.

rdwebdesign replied:
I understand the hack, but are you sure it's needed? 
You already increased the specificity with `body[class*="lcars"]`.

I will test again another time, but apparently it works without the hack.

MichalSvatos replied:
Nightly, with `lcars` or `lcars-picard` class. If I remove one of the `[class*="lcars"]` on line 398 - it breaks the checkboxes.

EDIT: Hmm, I see that maybe it doesn't have to be everywhere 🤔 For example the color, or the border. I wanted to be consistent 😄 I'll adjust that.

rdwebdesign replied:
@MichalSvatos

I finally found time to test your PR again.

When I use the simplified selector, I saw the issue only in "Domains" page (do you see it in other places?).

I noticed there are some things contributing for the issue, but we can fix them instead of adding a new hack or a lot of `!important` rules.

I pushed a new branch with the changes. ~~I will wait the team approval and them we can simplify the selectors here.~~ (merged)

~~The new code is in `development-v6`.~~ You can rebase your branch to use the new code.

EDIT:
The new code is in `devel`.

MichalSvatos replied:
Yes, it looks like that at the end only the "Domains" page cause an issue.

This PR is based on `devel` not `development-v6`, are you sure that I should rebase?

(I'm sorry for such a delay.)

rdwebdesign replied:
Sorry... it was a distraction when I typed.

New code is in **`devel`**.

PromoFaux commented on style/themes/lcars-picard.css:
```diff
@@ -0,0 +1,5882 @@
+
+/* --- additional visual tweaks */
+.navbar-custom-menu > .navbar-nav > li > .dropdown-menu::after {
+  content: "LCARS theme by @MichalSvatos";
```
@MichalSvatos - can you remove this please? No issue with the Author attribution in the header of the file.

The Pull Request / commits / release notes will provide credit

MichalSvatos replied:
Oh, sure! I've already removed one in the footer, I forgot about this one.

yubiuser commented:
I still think the position of the buttons at the audit log is kind of broken for long domains.
![Screenshot at 2023-10-25 23-16-33](https://github.com/pi-hole/web/assets/26622301/b219b835-e312-4d6f-b240-da2a622c0bed)

rdwebdesign commented:
You are right.
I probably didn't have long domains when I tested.

Something like this should fix:
 
```css
#domain-frequency .table-bordered > tbody > tr > td:last-child,
#ad-frequency .table-bordered > tbody > tr > td:last-child {
  width: 204px;
}
```

But I'm not sure if this will affect the ***Long-term Data > Top Lists***.

MichalSvatos commented:
I went through few iterations and this one was the good compromise between the functionality and least amount of code added. Other options were unecessaryli complex (eg. turning table rows in a flex containers). And simple `white-space: nowrap` rendered domains almost unreadable on the smaller screens. 

Not perfect solution, but it works. Anyway I open to suggestions.

yubiuser commented:
![Screenshot at 2023-11-01 14-14-59](https://github.com/pi-hole/web/assets/26622301/ae2f6526-9828-4d48-a8a4-49aa4086fc9f)


Looks good

yubiuser commented:
@MichalSvatos 

I think you could start working to adapt your theme to v6 now.....

MichalSvatos commented:
I tried to run `development-v6` in the docker the same way I did for `devel`, but it doesn't work 🙁

<details>
  <summary>My `docker_run.sh`</summary>
  
  ```
  #!/bin/bash

# https://github.com/pi-hole/docker-pi-hole/blob/master/README.md

PIHOLE_BASE="${PIHOLE_BASE:-$(pwd)}"
[[ -d "$PIHOLE_BASE" ]] || mkdir -p "$PIHOLE_BASE" || { echo "Couldn't create storage directory: $PIHOLE_BASE"; exit 1; }

# Note: FTLCONF_LOCAL_IPV4 should be replaced with your external ip.
docker run -d \
	--name pihole-v6 \
    -p 53:53/tcp -p 53:53/udp \
    -p 80:80 \
    -e TZ="America/Chicago" \
    --dns=127.0.0.1 --dns=1.1.1.1 \
    --restart=unless-stopped \
    --hostname pi.hole.v6 \
    -e VIRTUAL_HOST="pi.hole" \
    -e PROXY_LOCATION="pi.hole" \
    -e FTLCONF_LOCAL_IPV4="XXX.XXX.X.XX" \
    -e WEBPASSWORD="123456" \
    -e DNSMASQ_USER=root \
    -e PIHOLE_UID=501 \
    -e PIHOLE_GID=20 \
    pihole/pihole:development-v6

printf 'Starting up pihole container '
for i in $(seq 1 20); do
    if [ "$(docker inspect -f "{{.State.Health.Status}}" pihole)" == "healthy" ] ; then
        printf ' OK'
        echo -e "\n$(docker logs pihole 2> /dev/null | grep 'password:') for your pi-hole: http://${IP}/admin/"
        exit 0
    else
        sleep 3
        printf '.'
    fi

    if [ $i -eq 20 ] ; then
        echo -e "\nTimed out waiting for Pi-hole start, consult your container logs for more info (\`docker logs pihole\`)"
        exit 1
    fi
done;

  ```
</details>


<details>
  <summary>The error</summary>
  
  ```
  For security reasons, the password you type will not be visible.
  
  sudo: a terminal is required to read the password; either use the -S option to read from standard input or configure an askpass helper
  sudo: a password is required
  ```
  
</details>

rdwebdesign commented:
I never used this script and I never saw this error.
I use only a `docker run` command or a compose file.

**Note:**
The new image uses different Environment Variables (and their names are case sensitive).

You can find details in docker-pihole v6 readme:
https://github.com/pi-hole/docker-pi-hole/tree/development-v6

infinitytec commented:
Nice job!
